enhancements/baremetal/baremetal-operator-management.md

```diff
@@ -0,0 +1,166 @@
+---
+title: baremetal-operator-management
+authors:
+  - "@markmc"
+reviewers:
+  - "@abhinavdahiya"
+  - "@dhellmann"
+  - "@enxebre"
+  - "@eparis"
+  - "@hardys"
+  - "@sadasu"
+  - "@smarterclayton"
+  - "@stbenjam"
+approvers:
+  - TBD
+creation-date: 2020-02-13
+last-updated: 2020-02-13
+status: provisional
+see-also:
+  - https://github.com/openshift/enhancements/pull/200
+replaces:
+superseded-by:
+---
+
+# Managing the Bare Metal Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+1. Should the Bare Metal Operator be managed by the Cluster Version
+   Operator or the Machine API Operator?
+2. How to handle upgrades
+3. Which release this change should target
+
+## Summary
+
+The Bare Metal Operator provides bare metal machine management
```

Comment:
The use of "Bare Metal Operator" throughout this document is likely to be confusing. I think that we've all acknowledged that "baremetal-operator" was the wrong name for that component. It's really just the controller for BareMetalHost resources. It is one of several components that make up metal3, but it doesn't "operate" any other software in the sense that we usually mean when we talk about an operator.

What we're talking about is a new thing for managing the metal3 Deployment underpinning OpenShift. That new thing would be OpenShift-specific in a few ways, and likely not reusable for running metal3 as an application for deploying to hosts for purposes other than nodes in the current cluster because it would have to configure metal3 in a more constrained way than we would want it we were using it more generally (it can't allocate a PV for the Ironic database, it should only configure Ironic on 1 provisioning network, it must configure the host controller in the openshift-machine-api namespace, etc.).

Reply:
Fair points. I’m definitely assuming that the current baremetal-operator component will need to become more operator-like, and I've focusing this discussion on what should manage this evolved variant of baremetal-operator. I should be more explicit about that.

On naming … I included the link to metal3-io/baremetal-operator#227 in the references section for this reason, but again I could be more explicit about that. My tl;dr of that is "an operator's operand doesn't have to be software". It doesn't read to me like there's consensus that we shouldn't have called baremetal-operator an operator.


Anyway ... the current baremetal-operator image is just a binary that runs the control loop for the BareMetalHost resource. What concretely would we need to do to make BMO more operator-like? Something like this?

1. Decide what CVO stage (i.e. runlevel) BMO manifest should be applied. The easiest would be to stick with the 30-39 range reserved for the [Machine API](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)
2. Move the BareMetalHost and Provisioning CRDs (from the MAO) into the /manifests directory of the BMO image
3. (Optional?) Add a baremetal-operator ServiceAccount (rather than continuing to run it under machine-api-controllers)
4. (Optional?) Add an openshift-baremetal namespace (rather than continuing to use the openshift-machine-api namespace)
5. (Optional?) Add (or expand?) metrics support for BMO, and associated Service, ServiceMonitor and alerts.
6. Move the metal3 deployment from being dynamically generated by the MAO into a static manifest under /manifests in the BMO image.
7. Add a ClusterOperator resource and keep it updated - i.e. update its status field with the Available/Progressing/Degraded conditions, operator version number, and any relatedObjects
Tag the BMO image with io.openshift.release.operator=true

Certainly (6) looks like the most challenging, and perhaps this could be what triggers the need for an operator to manage that deployment with the Provisioning resource as the abstraction for driving that operator?

But ... point taken. I should make this assumption of "a future variant of BMO" clear, and we should have a separate enhancement to plan in detail what that operator would look like?

Reply:
> Anyway ... the current baremetal-operator image is just a binary that runs the control loop for the BareMetalHost resource. What concretely would we need to do to make BMO more operator-like? Something like this?

The thing we currently call the baremetal-operator doesn't even know it's provisioning nodes in a kubernetes cluster. I don't think we want to bake OpenShift-specific cluster management behavior into it, because that would make it less composable and likely kill any upstream interest.

> 1. Decide what CVO stage (i.e. runlevel) BMO manifest should be applied. The easiest would be to stick with the 30-39 range reserved for the [Machine API](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)
> 2. Move the BareMetalHost and Provisioning CRDs (from the MAO) into the /manifests directory of the BMO image

I'm not convinced that the Provisioning CRD is useful for anything other than configuring metal3 to run as part of the machine-api. It doesn't support the flexibility that I would expect from a general-purpose provisioning tool, like expressing how to scale the ironic deployment or how to use multiple provisioning networks. Neither of those are things we would want for a self-managing cluster, but that we would need for a single deployment of a centralized provisioning tool. Given that, I don't think we want it in the baremetal-operator repo. It should go into the repo with the new operator that will read it and create the Deployment for metal3 to serve the machine-api.

> 3. (Optional?) Add a baremetal-operator ServiceAccount (rather than continuing to run it under machine-api-controllers)

I'm not sure what that buys us for this case, but I don't think it would hurt?

> 4. (Optional?) Add an openshift-baremetal namespace (rather than continuing to use the openshift-machine-api namespace)

The cluster-api-provider-baremetal and remediation controller would have to be changed to look at the new namespace. It seems cleaner to just keep using the existing one.

> 5. (Optional?) Add (or expand?) metrics support for BMO, and associated Service, ServiceMonitor and alerts.
> 6. Move the metal3 deployment from being dynamically generated by the MAO into a static manifest under /manifests in the BMO image.

The Deployment is dynamically generated because we need to pass configuration values to containers that don't know how to read them directly from Kubernetes resources. Are you suggesting that we would make the BMO read those settings, then launch a separate Deployment with Ironic in it? That would break our model of using localhost to talk between services, which is a primary reason for having them all in the same Pod today.

> 7. Add a ClusterOperator resource and keep it updated - i.e. update its status field with the Available/Progressing/Degraded conditions, operator version number, and any relatedObjects
> 8. Tag the BMO image with io.openshift.release.operator=true
> 
> Certainly (6) looks like the most challenging, and perhaps this could be what triggers the need for an operator to manage that deployment with the Provisioning resource as the abstraction for driving that operator?

Yes, that approach makes a lot more sense. It avoids having a service being self-managing, which eliminates the issues with dynamically passing configuration into the services that don't know how to read it for themselves.

> But ... point taken. I should make this assumption of "a future variant of BMO" clear, and we should have a separate enhancement to plan in detail what that operator would look like?

I'm not sure it's a variant. It seems like it's just a new thing that acts in the ways you describe and launches the existing things. As David has pointed out elsewhere, we could forklift most of the existing MAO code that's doing it and drop it into a new repo.

Reply:
> > Anyway ... the current baremetal-operator image is just a binary that runs the control loop for the BareMetalHost resource. What concretely would we need to do to make BMO more operator-like? Something like this?
> 
> The thing we currently call the baremetal-operator doesn't even know it's provisioning nodes in a kubernetes cluster. I don't think we want to bake OpenShift-specific cluster management behavior into it, because that would make it less composable and likely kill any upstream interest.

Ok, another fair point taken - I'm entirely ignoring the complexity of having upstream, generically-applicable functionality and also OpenShift specific integration. (An easy mistake when we have an openshift/baremetal-operator fork that we could put the OpenShift specific stuff in!)

Would it be fair to simplify this as another point in favor of having a OpenShift specific operator managing an operand containing upstream, generically-applicable functionality?

> > 1. Decide what CVO stage (i.e. runlevel) BMO manifest should be applied. The easiest would be to stick with the 30-39 range reserved for the [Machine API](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/operators.md#how-do-i-get-added-as-a-special-run-level)
> > 2. Move the BareMetalHost and Provisioning CRDs (from the MAO) into the /manifests directory of the BMO image
> 
> I'm not convinced that the Provisioning CRD is useful for anything other than configuring metal3 to run as part of the machine-api. It doesn't support the flexibility that I would expect from a general-purpose provisioning tool, like expressing how to scale the ironic deployment or how to use multiple provisioning networks. Neither of those are things we would want for a self-managing cluster, but that we would need for a single deployment of a centralized provisioning tool. Given that, I don't think we want it in the baremetal-operator repo. It should go into the repo with the new operator that will read it and create the Deployment for metal3 to serve the machine-api.

Ok.

(This whole list ignored the existence of the upstream baremetal-operator project)

> > 1. (Optional?) Add a baremetal-operator ServiceAccount (rather than continuing to run it under machine-api-controllers)
> 
> I'm not sure what that buys us for this case, but I don't think it would hurt?
> 
> > 1. (Optional?) Add an openshift-baremetal namespace (rather than continuing to use the openshift-machine-api namespace)
> 
> The cluster-api-provider-baremetal and remediation controller would have to be changed to look at the new namespace. It seems cleaner to just keep using the existing one.

Both of these would be to fully separate the concerns of machine-api-operator (and its providers) from bare metal as a new "subsystem" (as Clayton describes it)

e.g. not having to change the machine-api-operator RBAC manifests to adjust the access to BMH resources in order to land a change to the bare metal components

(Exact details of what's needed/beneficial to be worked out)

> > 1. (Optional?) Add (or expand?) metrics support for BMO, and associated Service, ServiceMonitor and alerts.
> > 2. Move the metal3 deployment from being dynamically generated by the MAO into a static manifest under /manifests in the BMO image.
> 
> The Deployment is dynamically generated because we need to pass configuration values to containers that don't know how to read them directly from Kubernetes resources. Are you suggesting that we would make the BMO read those settings, then launch a separate Deployment with Ironic in it? That would break our model of using localhost to talk between services, which is a primary reason for having them all in the same Pod today.

No, nothing so radical - it's a hopeful "maybe we can make this Deployment static if we use things like configMapKeyRef to get these config values into env variables"

But, yes ... I know you just moved from a config map to this new Provisioning resource!

I guess I fully expect we'll decide the deployment has to be generated dynamically (as I said below), but I think it's worthwhile being clear on what exactly can't be done statically.

> > 1. Add a ClusterOperator resource and keep it updated - i.e. update its status field with the Available/Progressing/Degraded conditions, operator version number, and any relatedObjects
> > 2. Tag the BMO image with io.openshift.release.operator=true
> > 
> > Certainly (6) looks like the most challenging, and perhaps this could be what triggers the need for an operator to manage that deployment with the Provisioning resource as the abstraction for driving that operator?
> 
> Yes, that approach makes a lot more sense. It avoids having a service being self-managing, which eliminates the issues with dynamically passing configuration into the services that don't know how to read it for themselves.
> 
> > But ... point taken. I should make this assumption of "a future variant of BMO" clear, and we should have a separate enhancement to plan in detail what that operator would look like?
> 
> I'm not sure it's a variant. It seems like it's just a new thing that acts in the ways you describe and launches the existing things. As David has pointed out elsewhere, we could forklift most of the existing MAO code that's doing it and drop it into a new repo.

Ok, I'm seeing two points in favor of "a new thing":

1. baremetal-operator is an upstream project with a clearly defined scope that is generically applicable - we shouldn't evolve it into being an OpenShift specific thing

    Counterpoint - it could be a variant if the OpenShift fork includes the OpenShift integration. Or even as a new thing, it could just vendor the BMH controller code rather than launching baremetal-operator separately. (Not advocating strongly for either of those, just want to enumerate the possibilities)

2. The deployment to launch baremetal-operator (or a reasonable variant of it) could not be defined statically so that it could be included in the release image. So we need a new thing to define it.

Reply:
> Would it be fair to simplify this as another point in favor of having a OpenShift specific operator managing an operand containing upstream, generically-applicable functionality?

Yes. And it's specific to the OpenShift machine-api, not just OpenShift. Because a larger deployment for managing a bunch of non-cluster hosts would look quite different.

> > > 1. (Optional?) Add a baremetal-operator ServiceAccount (rather than continuing to run it under machine-api-controllers)
> > 
> > 
> > I'm not sure what that buys us for this case, but I don't think it would hurt?
> > > 1. (Optional?) Add an openshift-baremetal namespace (rather than continuing to use the openshift-machine-api namespace)
> > 
> > 
> > The cluster-api-provider-baremetal and remediation controller would have to be changed to look at the new namespace. It seems cleaner to just keep using the existing one.
> 
> Both of these would be to fully separate the concerns of machine-api-operator (and its providers) from bare metal as a new "subsystem" (as Clayton describes it)
>
> e.g. not having to change the machine-api-operator RBAC manifests to adjust the access to BMH resources in order to land a change to the bare metal components

I don't know the ramifications of the RBAC changes. I know the CAPBM needs to modify host CRs. It also watches them, so it would need to know about the new namespace. Which seems like it defeats any benefit we would get from moving to a new namespace.

The BMO doesn't need to modify Machine CRs. Maybe because we have that layering in place it's OK to use a different service account, but I would want to investigate more before committing to that change.

> (Exact details of what's needed/beneficial to be worked out)
> 
> > > 1. (Optional?) Add (or expand?) metrics support for BMO, and associated Service, ServiceMonitor and alerts.
> > > 2. Move the metal3 deployment from being dynamically generated by the MAO into a static manifest under /manifests in the BMO image.
> > 
> > 
> > The Deployment is dynamically generated because we need to pass configuration values to containers that don't know how to read them directly from Kubernetes resources. Are you suggesting that we would make the BMO read those settings, then launch a separate Deployment with Ironic in it? That would break our model of using localhost to talk between services, which is a primary reason for having them all in the same Pod today.
> 
> No, nothing so radical - it's a hopeful "maybe we can make this Deployment static if we use things like configMapKeyRef to get these config values into env variables"
> 
> But, yes ... I know you just moved from a config map to this new Provisioning resource!
> 
> I guess I fully expect we'll decide the deployment has to be generated dynamically (as I said below), but I think it's worthwhile being clear on what exactly can't be done statically.

Something needs to convert the Provisioning CR into values to give to the Deployment, because the things running inside the Pod created by the Deployment can't read the CR themselves. We could use a ConfigMap, but we would have to generate that dynamically because we're not allowed to create it from the installer. So it seems like we might as well just create the Deployment dynamically, instead. That gives us the benefit of eventually being able to change some of the values at runtime, if we decide we need to.

> > > 1. Add a ClusterOperator resource and keep it updated - i.e. update its status field with the Available/Progressing/Degraded conditions, operator version number, and any relatedObjects
> > > 2. Tag the BMO image with io.openshift.release.operator=true
> > > 
> > > Certainly (6) looks like the most challenging, and perhaps this could be what triggers the need for an operator to manage that deployment with the Provisioning resource as the abstraction for driving that operator?
> > 
> > 
> > Yes, that approach makes a lot more sense. It avoids having a service being self-managing, which eliminates the issues with dynamically passing configuration into the services that don't know how to read it for themselves.
> > > But ... point taken. I should make this assumption of "a future variant of BMO" clear, and we should have a separate enhancement to plan in detail what that operator would look like?

I'm not sure where we would want to describe it. I'd like to see it in the same place where we agree to the move, so we're detailing the entire plan in one place. Maybe that's here?

> > I'm not sure it's a variant. It seems like it's just a new thing that acts in the ways you describe and launches the existing things. As David has pointed out elsewhere, we could forklift most of the existing MAO code that's doing it and drop it into a new repo.
> 
> Ok, I'm seeing two points in favor of "a new thing":
> 
> 1. baremetal-operator is an upstream project with a clearly defined scope that is generically applicable - we shouldn't evolve it into being an OpenShift specific thing
>    Counterpoint - it could be a variant if the OpenShift fork includes the OpenShift integration. Or even as a new thing, it could just vendor the BMH controller code rather than launching baremetal-operator separately. (Not advocating strongly for either of those, just want to enumerate the possibilities)
> 2. The deployment to launch baremetal-operator (or a reasonable variant of it) could not be defined statically so that it could be included in the release image. So we need a new thing to define it.

Point 2 means that the counter point to point 1 won't work.

Reply:
Ok, the new revision and the prototype implementation is heavily influence by this discussion. Would greatly appreciate your review. Thanks.

One thing that's not addressed (except for the "upstream implications" FIXME) is whether upstream Metal3 should have an operator like this, but minus the OpenShift specific stuff and with support for use cases beyond the OpenShift use case. Not forgotten, it's definitely a TODO item!